src/map_ref.rs

```diff
@@ -0,0 +1,272 @@
+use crate::iter::*;
+use crate::HashMap;
+use crossbeam_epoch::Guard;
+use std::borrow::Borrow;
+use std::fmt::{self, Debug, Formatter};
+use std::hash::{BuildHasher, Hash};
+use std::ops::{Deref, Index};
+
+/// A reference to a [`HashMap`], constructed with [`HashMap::guarded`] or [`HashMap::with_guard`].
+///
+/// The current thread will be pinned for the duration of this reference.
+/// Keep in mind that this prevents the collection of garbage generated by the map.
+pub struct HashMapRef<'map, K: 'static, V: 'static, S = crate::DefaultHashBuilder> {
+    map: &'map HashMap<K, V, S>,
+    guard: GuardRef<'map>,
```

Comment:
I think we may also want to have
```rust
pub fn repin(&mut self)
```
Which calls [`Guard::repin`](https://docs.rs/crossbeam-epoch/0.8.0/crossbeam_epoch/struct.Guard.html#method.repin).

Reply:
We can do that in the `Owned` case, but we don't have exclusive access to the guard in the `Ref` case from `with_guard`. Would you have `repin` be a no-op in that case? Or should we return a `bool` whether we were able to actually call `repin`?

Or we could just expose the guard with a couple of methods, so you can `repin` it yourself.

```rust
    fn guard(&self) -> &Guard;
    fn guard_mut(&mut self) -> Option<&mut Guard>;
```

e.g. `map_ref.guard_mut().map(Guard::repin)`

Reply:
Ah, damn, you're right...Hmm. Maybe the `repin` use case is specific enough that if you want that behavior you have to use the non-`HashMapRef` interface. That seems fine. The methods you gave above would work, but I think in practice they'd probably go essentially unused.